### ISSUE TITLE
Reset button on iOS time input does not return correct event value

### ISSUE THREAD

**Issue by PingEleanor1** [facebook/react#23299]
React version: React 17.0.2
IOS version: 14+

## Bug
For iOS(Safari and Chrom).
The reset button function does not work on the time input:

 If I change time to 00:31, reset directly, the time jumps back to 00:30, but the computer still shows 00:31.
When pressing reset on a time input, the onChange event is not fired.

![reset](https://user-images.githubusercontent.com/83808682/154050608-748e912a-c6c7-4cae-9d27-9a705762b6d2.PNG)

Link to code example:
https://stackblitz.com/edit/react-enuflz?file=src%2FApp.js
Pls try on iOS.

**Comment by sammy-SC** [comment]
Hello @PingEleanor1,

thank you for the report. I was able to repro the issue and

**Comment by mmarkelov** [comment]
Seems like it's related to https://github.com/facebook/react/issues/8938

**Comment by shriphad-rao** [comment]
Is this issue still open? I'm interested in working on it and would love to contribute.

**Comment by BikramjeetSingh** [comment]
Yes, this issue is still open, and it is affecting some of my code right now. React v17.0.2, iPhone XR, iOS 16.6

**Comment by BartoszDadok** [comment]
@shriphad-rao Have you found any solution?

**Comment by alexandremacedo** [comment]
Workaround that I use

`
const dateRef = useRef()

function handleDate(event) {
    const target = event.target;
    setTimeout(() => {
      target.defaultValue = "";
    }, 100);
}

<input ref={dateRef}  onChange={handleDate} value={state}/>
`

**Comment by nikischin** [comment]
> Workaround
> 
> ```
> const dateRef = useRef()
> 
> function handleDate(event) {
>     const target = event.target;
> 
>     setTimeout(() => {
>       target.defaultValue = "";
>     }, 100);
> }
> 
> <input ref={dateRef}  onChange={handleDate} value={state}/>```
> 

What is the `ref` needed for in this example? It seems like it was set but never used. Kinda like a mix from the examples provided in #8938

**Comment by nikischin** [comment]
> > > Workaround
> > > ```
> > > const dateRef = useRef()
> > > 
> > > function handleDate(event) {
> > >     const target = event.target;
> > > 
> > >     setTimeout(() => {
> > >       target.defaultValue = "";
> > >     }, 100);
> > > }
> > > 
> > > <input ref={dateRef} onChange={handleDate} value={state}/>
> > > ```
> > 
> > 
> > What is the `ref` needed for in this example? It seems like it was set but never used. Kinda like a mix from the examples provided in #8938
> > Really don't understand why this issue is open for so long already and still not fixed.
> 
> I found a working workaround. I've used nativeEvent directly. Though Typescript support is not present for this so I've added a fallback to the react event in case native event is not present (by TS definition)
> 
> ```tsx
> const onDateChange = (e: React.ChangeEvent<HTMLInputElement>) => {
>     const date = new Date(e.nativeEvent.target && 'value' in e.nativeEvent.target && typeof e.nativeEvent.target.value === 'string' ? e.nativeEvent?.target?.value : e.target.value);
> 
>     if (isValidDate(date)) {
>       setDate(date);
>     } else {
>       setDate(null);
>     }
>   };
> 
> return  <input onChange={onDateChange} />
> ```

You are missing the implementation of isValidDate. Assuming an implementation like this it unfortunately does not fully work:

```
function isValidDate(d: Date) {
    return d instanceof Date && !isNaN(d);
}
```

Unfortunately for me the click on Reset doesn't reliable trigger the change Event. It is triggered whenever, I opened the dialog and changed the date and press reset before closing the dialog. However, it does not work, if I open the dialog with a date already set and press on reset without changing the date before.

**Comment by LeSingh1** [comment]
Hi, I’m going to try reproducing this on a current React build with iOS Safari. If I can still reproduce it, I’ll investigate the React DOM input/change event path and share a focused fix or findings.

**Comment by LeSingh1** [comment]
Followed up on this with a code-level investigation against current `main`. Sharing findings in case it helps move this forward.

**Event path:** `change` on `<input type="time">` flows through `ChangeEventPlugin.extractEvents` → `isTextInputElement` (✅ `time` is in the supported list) → `getTargetInstForInputOrChangeEvent` → `getInstIfValueChanged` → `updateValueIfChanged` in `inputValueTracking.js`. The gate is the tracker-vs-`node.value` comparison: equal → onChange suppressed. There is no `type === 'time'` early-return — the generic path handles it.

**Why the bug surfaces on iOS:** When a `change` event arrives at React, `updateValueIfChanged` reads `node.value` via the original property getter. On iOS Safari's native Reset, the JS-visible `node.value` does not synchronously reflect the reset at change-event dispatch time. Two failure modes follow:

1. `node.value` still reads the prior value → tracker == node → onChange suppressed entirely (matches @nikischin's "doesn't reliably trigger" report).
2. `node.value` reads `""` → onChange fires, but `e.target.value` may revert by the time userland reads it due to React's `enqueueStateRestore` writing the controlled value back (matches @WinglessFrame's "nativeEvent has correct value, e.target is stale" observation).

**Why a clean React-side fix is hard:** A minimal patch could bypass `updateValueIfChanged` for `change` events on date/time inputs, forcing onChange to dispatch. But this doesn't solve the user-visible problem — if Safari hasn't propagated the reset value to `node.value` at event-dispatch time, the handler still receives the stale value and updates state to the same value, so the controlled input never empties.

Without a way to synchronously obtain the post-reset value from Safari, this appears to be primarily a WebKit timing issue rather than something React can fully fix synchronously. The userland workarounds in this thread (deferred `defaultValue = ""` via `setTimeout`, reading `e.nativeEvent.target.value`) work around the timing, not the root cause.

Would maintainers want a focused React-side mitigation, such as bypassing the value-change gate for `change` on date/time inputs with a JSDOM test, despite the limitation that it cannot fully resolve the stale-value case? Happy to follow up in that direction if maintainers think it would be useful.
